bob-backend/src/pearl/group.rs

```diff
@@ -569,39 +598,79 @@ impl Group {
         });
     }
 
-    pub async fn delete(&self, key: BobKey) -> Result<u64, Error> {
-        let holders = self.holders.read().await;
+    pub async fn delete(
+        &self,
+        key: BobKey,
+        is_alien: bool,
+        timestamp_config: StartTimestampConfig,
+        force_delete: bool,
+    ) -> Result<u64, Error> {
+        let is_empty = self.holders.read().await.len() == 0;
+        // We keep holders unnlocked because actual holder should created exactly one holder anyway
+        if is_empty {
+            let holder = self
+                .get_actual_holder(get_current_timestamp(), timestamp_config)
+                .await?;
+            self.delete_in_holders(
+                std::iter::once((holder.0, &holder.1)),
+                key,
+                is_alien,
+                force_delete,
+            )
+            .await
+        } else {
+            let holders = self.holders.read().await;
+            self.delete_in_holders(holders.iter().enumerate(), key, is_alien, force_delete)
+                .await
+        }
+    }
+
+    async fn delete_in_holders(
+        &self,
+        holders: impl Iterator<Item = (ChildId, &Holder)>,
+        key: BobKey,
+        is_alien: bool,
+        force_delete: bool,
+    ) -> Result<u64, Error> {
         let mut total_count = 0;
-        for holder in holders.iter() {
-            let delete = Self::delete_common(holder.clone(), key).await;
+        for holder in holders {
+            let delete = Self::delete_common(holder.1.clone(), key, force_delete).await;
+            if is_alien {
```

Comment:
Logic obviously incorrect. You have the `is_alien` and `force_delete` flags, which overlaps with each over. Looks like you should use only `force_delete` and remove `is_alien`